x-pack/platform/plugins/shared/cases/public/components/markdown_editor/paste_upload/use_image_paste_upload.test.tsx

```diff
@@ -0,0 +1,274 @@
+        () =>
+          useImagePasteUpload({
+            editorRef,
+            field,
+            caseId: CASE_ID,
+            owner: OWNER,
+            fileKindId: FILE_KIND_ID,
+          }),
+        { wrapper }
+      );
+
+      return { textarea, field, result };
+    };
+
+    it('returns initial uploading state false', () => {
+      const { result } = setup();
+      expect(result.current.isUploading).toBe(false);
+    });
+
+    it('starts upload when an image is pasted', async () => {
+      const { textarea, result } = setup();
+
+      const imageFile = new File(['x'], 'image.png', { type: 'image/png' });
+      const dt = {
+        items: [
+          {
+            kind: 'file',
+            type: imageFile.type,
+            getAsFile: () => imageFile,
+          },
+        ],
+        types: ['Files'],
+      };
+
+      act(() => {
+        // @ts-expect-error partial impl
+        textarea.dispatchEvent(new ClipboardEvent('paste', { clipboardData: dt, bubbles: true }));
+
+        // Emit files$ to trigger START_UPLOAD placeholder insertion
+        mockUploadState.files$.next([
+          { file: imageFile, status: 'uploading' } as unknown as FileState,
+        ]);
+
+        // simulate UploadState emitting that upload is in progress
+        mockUploadState.uploading$.next(true);
+      });
+
+      // wait until hook reflects uploading state
+      await waitFor(() => expect(result.current.isUploading).toBe(true));
+    });
+
+    it('sets error when unsupported mime is pasted', () => {
+      const { textarea, result } = setup();
+
+      const textFile = new File(['x'], 'file.txt', { type: 'text/plain' });
+      const dt = {
+        items: [
+          {
+            kind: 'file',
+            type: textFile.type,
+            getAsFile: () => textFile,
+          },
+        ],
+        types: ['Files'],
+      };
+
+      act(() => {
+        // @ts-expect-error partial implementation for testing
+        textarea.dispatchEvent(new ClipboardEvent('paste', { clipboardData: dt, bubbles: true }));
+      });
+
+      // error should be propagated through reducer state
+      expect(result.current.errors).toEqual([UNSUPPORTED_MIME_TYPE_MESSAGE]);
+    });
+
+    it('sets error when multiple files are pasted simultaneously', () => {
+      const { textarea, result } = setup();
+      const file1 = new File(['x'], 'image1.png', { type: 'image/png' });
+      const file2 = new File(['y'], 'image2.png', { type: 'image/png' });
+      const dt = {
+        items: [
+          { kind: 'file', type: file1.type, getAsFile: () => file1 },
+          { kind: 'file', type: file2.type, getAsFile: () => file2 },
+        ],
+        types: ['Files'],
+      };
+      act(() => {
+        // @ts-expect-error partial impl
+        textarea.dispatchEvent(new ClipboardEvent('paste', { clipboardData: dt, bubbles: true }));
+      });
+
+      expect(result.current.errors).toEqual([NO_SIMULTANEOUS_UPLOADS_MESSAGE]);
```

Comment:
I didn't see an error in the UI, I just couldn't paste anything else in while the upload was happening. Is this expected or is the error supposed to propagate back to the UI? Personally I like not having an in your face UI error, but is there an error happening in the reducer behind the scenes?

Reply:
I think there might be a misunderstanding about the intention here.

Yes, if you have an upload in progress, we block a second one from kicking off. I.e., I paste an image and while that upload is in progress I screenshot again and paste that. In this case, the UI indicates the upload is happening and I think it's reasonable to expect the user to realize their action was blocked. If they try again when the upload finishes, they'll get a new one.

This is intended for cases where the user sends multiple files **as part of a single paste event**. Like if you highlight 3 image files in your desktop, put them in your clipboard, and paste them here. You should expect to see the error message in that case.